cimbar/cimbar.py

```diff
@@ -114,41 +119,76 @@ def _get_decoder_stream(outfile, ecc, fountain):
     return reed_solomon_stream(f, ecc, mode='write', on_failure=on_rss_failure) if ecc else f
 
 
-def decode_iter(src_image, dark, force_preprocess, deskew, auto_dewarp):
-    should_preprocess = force_preprocess
+def compute_tint(img, dark):
+    def update(c, r, g, b):
+        c['r'] = max(c['r'], r)
+        c['g'] = max(c['g'], g)
+        c['b'] = max(c['b'], b)
+
+    cc = {}
+    cc['r'] = cc['g'] = cc['b'] = 1
+
+    if dark:
+        pos = [(28, 28), (28, 992), (992, 28)]
+    else:
+        pos = [(67, 0), (0, 67), (945, 0), (0, 945)]
+
+    for x, y in pos:
+        iblock = img.crop((x, y, x + 4, y + 4))
+        r, g, b = avg_color(iblock)
+        update(cc, *avg_color(iblock))
+
+    print(f'tint is {cc}')
+    return cc['r'], cc['g'], cc['b']
+
+
+def _decode_iter(ct, img, color_img):
+    cell_pos = cell_positions(CELL_SPACING, CELL_DIMENSIONS, CELLS_OFFSET)
+    finder = AdjacentCellFinder(cell_pos, CELL_DIMENSIONS)
+    decode_order = FloodDecodeOrder(cell_pos, finder)
+    for i, (x, y), drift in decode_order:
+        best_bits, best_dx, best_dy, best_distance = _decode_cell(ct, img, color_img, x, y, drift)
+        decode_order.update(best_dx, best_dy, best_distance)
+        yield i, best_bits
+
+
+def decode_iter(src_image, dark, should_preprocess, should_color_correct, deskew, auto_dewarp):
     tempdir = None
     if deskew:
         tempdir = TemporaryDirectory()
         temp_img = path.join(tempdir.name, path.basename(src_image))
         dims = detect_and_deskew(src_image, temp_img, dark, auto_dewarp)
-        should_preprocess |= dims[0] < TOTAL_SIZE or dims[1] < TOTAL_SIZE
+        if should_preprocess < 0:
+            should_preprocess = dims[0] < TOTAL_SIZE or dims[1] < TOTAL_SIZE
         color_img = Image.open(temp_img)
     else:
         color_img = Image.open(src_image)
+
     ct = CimbDecoder(dark, symbol_bits=BITS_PER_SYMBOL, color_bits=BITS_PER_COLOR)
     img = _preprocess_for_decode(color_img) if should_preprocess else color_img
 
-    cell_pos = cell_positions(CELL_SPACING, CELL_DIMENSIONS, CELLS_OFFSET)
-    finder = AdjacentCellFinder(cell_pos, CELL_DIMENSIONS)
-    decode_order = FloodDecodeOrder(cell_pos, finder)
-    for i, (x, y), drift in decode_order:
-        best_bits, best_dx, best_dy, best_distance = _decode_cell(ct, img, color_img, x, y, drift)
-        decode_order.update(best_dx, best_dy, best_distance)
-        yield i, best_bits
+    if should_color_correct:
+        from colormath.chromatic_adaptation import _get_adaptation_matrix
```

Comment:
Not sure about whether `colormath` should the dependency, or `colour`. I'm leaning towards the latter for more extensive color tricks, but for the simple adaptation matrix check this works.

(could also just pull the code out of colormath/wherever, there's not a lot to it)